### PR TITLE
streamline terminal input

### PR DESCRIPTION
**Pull request by maxtaco**
- use ssh/terminal for everything, but in paticular my patch, which correctly cancels out on Control-C
- get rid of miniline
- get rid of gopass

### CODE REVIEW

**Comment by maxtaco** [comment]
could use a bunch of thumbs on this one....

@oconnor663 / @patrickxb / @zanderz: what do you think on linux / osx / windows?

File: go/vendor/github.com/keybase/go-crypto/ssh/terminal/terminal.go [hunk 1]
```diff
@@ -0,0 +1,899 @@
...
+// that the user has entered.
+func (t *Terminal) handleKey(key rune) (line string, ok bool, err error) {
+	if t.pasteActive && key != keyEnter {
+		t.addKeyToLine(key)
+		return
+	}
+
+	switch key {
+	case keyCtrlC:
+		err = io.ErrUnexpectedEOF
```
Comment (maxtaco) [inline go/vendor/github.com/keybase/go-crypto/ssh/terminal/terminal.go]: this is my patch to go's library...

File: go/vendor/github.com/keybase/go-crypto/ssh/terminal/terminal.go [hunk 2]
```diff
@@ -0,0 +1,899 @@
...
+			} else if key == keyPasteEnd {
+				t.pasteActive = false
+				continue
+			}
+			if !t.pasteActive {
+				lineIsPasted = false
+			}
+			line, lineOk, err = t.handleKey(key)
+			if err != nil {
+				return line, err
```
Comment (maxtaco) [inline go/vendor/github.com/keybase/go-crypto/ssh/terminal/terminal.go]: also propagate the error back out....

**Comment by zanderz** [comment]
I see Windows support in there, looks good. I will have to try the sample.
Curious what happens with color codes: they are supported in the Windows 10 terminal but not before.

**Comment by maxtaco** [comment]
Yeah that terminal code is already in use elsewhere in the code base. I
just forked and vendored it for the control c feature.

**Comment by patrickxb** [comment]
This looks good, I just want to make sure that we're not losing anything that we previously used miniline/gopass for.

I believe the need for miniline was solely for ctrl-c:  https://github.com/keybase/client/pull/933 which it looks like this handles.

And gopass was for this?  https://github.com/keybase/client/pull/1637

**Comment by maxtaco** [comment]
It seems like the ssh terminal always did the right thing here but good
point! @oconnor663 what do you think?

**Comment by oconnor663** [comment]
LGTM. I've always been skeptical of us handing ctrl-c as an EOF instead of a signal (possibly because I deeply misunderstand how any of this terminal baloney works ^.^), but it looks like this is just refactoring things so that we don't need as many dependencies to do what we were already doing?

**Comment by maxtaco** [comment]
Yes.  And I had to had control-c support into ssh terminal which didn't
previously exist.

File: go/minterm/minterm.go [hunk 3]
```diff
@@ -42,7 +42,7 @@ func (m *MinTerm) Shutdown() error {
 	if m.termIn != nil {
 		go m.termIn.Close()
 	}
-	if m.termOut != nil {
+	if m.termOut != nil && m.closeTermOut {
```
Comment (oconnor663) [inline go/minterm/minterm.go]: Ah sorry I missed this the first time through. Why was this working before? Or was it silently risking a close-someone-else's-file condition?

File: go/minterm/minterm.go [hunk 4]
```diff
@@ -76,12 +73,38 @@ func (m *MinTerm) PromptPassword(prompt string) (string, error) {
...
 	if err != nil {
-		if err == gopass.ErrInterrupted {
-			err = ErrPromptInterrupted
-		}
 		return "", err
 	}
-	return string(b), nil
+	defer terminal.Restore(fd, oldState)
+	var ret string
+	ret, err = terminal.NewTerminal(m.termIn, "").ReadPassword("")
```
Comment (oconnor663) [inline go/minterm/minterm.go]: Looks like we're doing the right thing with respect to #1637. Thanks for remembering that, @patrickxb!

File: go/minterm/minterm.go (hunk 3 above)
Comment (maxtaco) [inline go/minterm/minterm.go]: I'm not sure, but it was definitely double-closing!

**Pull request merged**